File: alkash3d-luascript/examples/bobber.lua
-- bobber.lua — эталонный пример .lua-скрипта для alkash3d-luascript.
--
-- Соглашение по протоколу (см. комментарий в src/lib.rs):
--   update(dt, x, y, z, rx, ry, rz) -> (new_x, new_y, new_z, changed)
--     вызывается каждый кадр; changed=1 (число), если позицию нужно
--     применить, иначе 0 — движок оставит Transform как есть.
--   on_event(event_type, data0, data1, data2, data3)
--     вызывается по событию (см. ScriptEventType в scripting_api.rs:
--     0=Custom, 1=CollisionEnter, 2=CollisionExit, 3=ZoneEnter,
--     4=ZoneExit, 5=Spawned, 6=Despawned).
--
-- Демонстрирует ровно то же самое, что и Rust-пример
-- (alkash3d-examplescript) — покадровое покачивание вверх-вниз по
-- синусоиде + изменение амплитуды по событию ZoneEnter/Custom, чтобы
-- было наглядно видно: один и тот же сценарий работает одинаково и на
-- нативном (C++/Rust), и на Lua языке скриптинга.

local base_x, base_y, base_z = nil, nil, nil
local time_accum = 0.0
local amplitude = 0.3    -- метры
local bob_speed = 2.0    -- рад/сек

function update(dt, x, y, z, rx, ry, rz)
    -- Точка покоя фиксируется один раз, при первом кадре — дальше
    -- покачивание идёт вокруг НЕЁ, а не вокруг текущей позиции каждого
    -- кадра (иначе амплитуда накапливалась бы, а не колебалась).
    if base_x == nil then
        base_x, base_y, base_z = x, y, z
    end

    time_accum = time_accum + dt
    local offset_y = math.sin(time_accum * bob_speed) * amplitude

    return base_x, base_y + offset_y, base_z, 1
end

function on_event(event_type, data0, data1, data2, data3)
    -- ZoneEnter (3) или Custom (0): data0 > 0 — новый множитель амплитуды
    -- относительно базовой (0.3) — то же соглашение, что и в
    -- Rust-примере, чтобы оба скрипта были взаимозаменяемы для теста.
    if event_type == 3 or event_type == 0 then
        if data0 > 0.0 then
            amplitude = 0.3 * data0
        end
    end
end

File: alkash3d-rust/examples_scripts/bobber.py
# bobber.py — эталонный пример .py-скрипта для встроенного Python
# hot-reload интерпретатора движка (см. engine/scripting_python.rs).
#
# Соглашение по протоколу — ТО ЖЕ САМОЕ, что и у Lua (bobber.lua в
# alkash3d-luascript/examples/), намеренно, чтобы один и тот же сценарий
# был взаимозаменяем между всеми языками скриптинга:
#   update(dt, x, y, z, rx, ry, rz) -> (new_x, new_y, new_z, changed)
#     вызывается каждый кадр; changed=1, если позицию нужно применить.
#   on_event(event_type, data0, data1, data2, data3)
#     вызывается по событию (см. ScriptEventType в scripting_api.rs).
#
# Особенность именно Python-варианта: этот файл можно редактировать ПРЯМО
# во время работы движка — при следующем кадре движок заметит изменение
# mtime файла и автоматически перезагрузит его (см.
# PythonScriptRuntime::check_hot_reload) — без перезапуска игры.

import math

_base = None
_time_accum = 0.0
_amplitude = 0.3   # метры
_bob_speed = 2.0   # рад/сек


def update(dt, x, y, z, rx, ry, rz):
    global _base, _time_accum

    # Точка покоя фиксируется один раз, при первом кадре.
    if _base is None:
        _base = (x, y, z)

    _time_accum += dt
    offset_y = math.sin(_time_accum * _bob_speed) * _amplitude

    return (_base[0], _base[1] + offset_y, _base[2], 1)


def on_event(event_type, data0, data1, data2, data3):
    global _amplitude

    # ZoneEnter (3) или Custom (0): data0 > 0 — новый множитель амплитуды.
    if event_type == 3 or event_type == 0:
        if data0 > 0.0:
            _amplitude = 0.3 * data0
